src/backend/distributed/connection/placement_connection.c

```diff
@@ -837,7 +837,18 @@ CheckShardPlacements(ConnectionShardHashEntry *shardEntry,
 
 		if (placementEntry->failed)
 		{
-			UpdateShardPlacementState(placementEntry->key.placementId, FILE_INACTIVE);
+			uint64 shardId = shardEntry->key.shardId;
+			uint64 placementId = placementEntry->key.placementId;
+			ShardPlacement *shardPlacement = LoadShardPlacement(shardId, placementId);
+
+			/*
+			 * We only set shard state if its current state is FILE_FINALIZED, which
+			 * prevents overwriting shard state if it is already set at somewhere else.
+			 */
+			if (shardPlacement->shardState == FILE_FINALIZED)
+			{
+				UpdateShardPlacementState(placementEntry->key.placementId, FILE_INACTIVE);
+			}
 		}
```

Comment:
It makes sense to put this in a separate commit. I did that as you suggested.

This is related with this PR because if we cannot establish a connection and error occurs during RemoteTransactionBeginIfNecessary(), we set shard state of related shards to FILE_TO_DELETE, but then connection management sets their shard state to FILE_INACTIVE. To reproduce this case, just stop the postgres instance at one of the workers then DROP TABLE from the master.